libafl/src/executors/inprocess/inner.rs

```diff
@@ -65,36 +66,62 @@ impl<HT, I, OT, S> HasObservers for GenericInProcessExecutorInner<HT, I, OT, S>
     }
 }
 
-impl<HT, I, OT, S> GenericInProcessExecutorInner<HT, I, OT, S>
-where
-    OT: ObserversTuple<I, S>,
+/// A simple guard, to be used with [`GenericInProcessExecutorInner`].
+#[derive(Debug)]
+pub struct InProcessGuard<'a> {
+    phantom: PhantomData<&'a ()>,
+}
+
+impl Drop for InProcessGuard<'_> {
+    fn drop(&mut self) {
+        unsafe {
+            let data = &raw mut GLOBAL_STATE;
+
+            write_volatile(&raw mut (*data).current_input_ptr, null());
+            compiler_fence(Ordering::SeqCst);
+        }
+    }
+}
+
+impl<E, EM, HT, I, OT, S, Z> EntersTarget<E, EM, I, S, Z>
+    for GenericInProcessExecutorInner<HT, I, OT, S>
 {
+    type Guard<'a>
+        = InProcessGuard<'a>
+    where
+        E: 'a,
+        EM: 'a,
+        I: 'a,
+        S: 'a,
+        Z: 'a;
+
     /// This function marks the boundary between the fuzzer and the target
     ///
     /// # Safety
     /// This function sets a bunch of raw pointers in global variables, reused in other parts of
     /// the code.
-    // TODO: Remove EM and Z from function bound and add it to struct instead to avoid possible type confusion
     #[inline]
-    pub unsafe fn enter_target<EM, Z>(
-        &mut self,
-        fuzzer: &mut Z,
-        state: &mut S,
-        mgr: &mut EM,
-        input: &I,
-        executor_ptr: *const c_void,
-    ) {
+    fn enter_target<'a>(
+        executor: &'a mut E,
+        fuzzer: &'a mut Z,
+        state: &'a mut S,
+        mgr: &'a mut EM,
+        input: &'a I,
+    ) -> Self::Guard<'a> {
         unsafe {
             let data = &raw mut GLOBAL_STATE;
```

Comment:
Add a `read_volatile` check here that `current_input_ptr` is `ptr::null()` please

And add the potential race condition to the # Safety comment

Reply:
Correct me if I'm wrong, but I don't think there's a race threat here unless there are multiple executors running simultaneously. If there is, then signal handling is already broken, since this would apply at a thread group level.

Reply:
Yes if you call this function twice you have a race condition. Ergo, document.
Also, return Err or None if it's already set

Reply:
I'm not saying it's how we use it, but it's how the doc should be

Reply:
Ah but you can't call this function twice, because you'll hold the guard.

Reply:
And, if you're not holding the guard, then the guard is dropped and it's already restored the guarded state.